[FLINK-12146][network] Remove unregister task from NetworkEnvironment to simplify the interface of ShuffleService

zhijiangW opened the pull request:
## What is the purpose of the change

*`NetworkEnvironment#unregisterTask` is used for closing partition/gate and releasing partition from `ResultPartitionManager`. partition/gate close could be done in task which already maintains the arrays of them. Further we could release partition from `ResultPartitionManager` inside `ResultPartition` via introducing `ResultPartition#fail(Throwable)`. To do so, the `NetworkEnvironment#unregisterTask` could be totally replaced to remove. The benefit is simplifying the method of `NetworkEnvironment` which would be regarded as default `ShuffleService` implementation.*

## Brief change log

  - *Remove `unregisterTask` from `NetworkEnvironment`*
  - *Introduce `close(Throwable)` in `ResultPartition` for releasing*
  - *Fix the related logic in task class*

## Verifying this change

*This change is already covered by existing tests*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhijiangW commented:
@azagrebin I ever thought that unifying the close of partition/gate with canceler process, but it seems a bit different during handling exception, so just make them independent as now.

In addition, I am not sure why the previous implementation only catches exception for gate close, and the partition close might also cause exception. So I just refactored the processes and kept the behavior as before to not change anything.

azagrebin commented:
@flinkbot approve all

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -879,6 +875,24 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
+
+		for (ResultPartition partition : producedPartitions) {
+			taskEventDispatcher.unregisterPartition(partition.getPartitionId());
+			partition.close();
```
but I think it would be ok, if we actually add one more method `closeNetworkResources` and put there partition/gate closings from `TaskCanceler.run` and use here in `releaseNetworkResources` after `taskEventDispatcher.unregisterPartition` and `if () partition.fail` loops. We will eliminate code duplication and improve log/exception handling in former `unregisterTask`.

zhijiangW replied:
If you also think so, we could integrate the network release with the process in task canceler. :)

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -326,6 +326,23 @@ public void release(Throwable cause) {
+	 *
+	 * @param throwable which indicates FAILED/CANCELING task if not null
+	 */
+	public void close(Throwable throwable) {
```
That is a good idea. The semantic is more clearly.

zhijiangW commented:
@azagrebin  thanks for reviews and the good inline suggestions!

I was supposed to submit a separate fixup commit for addressing your comments.  But when I ament the first commit message to add more descriptions, the new code modifications has squashed with previous commit automatically.  Considering the small overall changes which might not bring trouble for your further review, I keep the current status. :)

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -879,6 +875,42 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
+	 * There are two scenarios to release the network resources. One is from {@link TaskCanceler} to early
+	 * release partitions and gates. Another is from task thread during task exiting.
+	 */
+	private void closeNetworkResources(boolean isCanceling) {
```
I would break this method down:
```
private void releaseNetworkResources() {
  for all partitions taskEventDispatcher.unregisterPartition
  Throwable cause = isCanceledOrFailed() ? getFailureCause() : null;
  closeOrFailNetworkResources(producedPartitions, inputGates, cause);
}

static closeNetworkResources(producedPartitions, inputGates, cause) {
  // closing loops or fail for partition if cause is null
}

TaskCanceler.run() {
  closeOrFailNetworkResources(producedPartitions, inputGates, null); // to preserve what we have
}
```
I would also keep TaskCanceler class static to simplify refactoring in future.

zhijiangW replied:
I agree with the option of keeping previous static TaskCanceler class. 
But the static `closeNetworkResources` might need two more parameters. One is for `taskNameWithSubtask` used for log, and another is the boolean `isCanceledOrFailed` because in previous behavior the result of  `isCanceledOrFailed` is not always equal to `cause != null`. In the case of `Task#cancelExecution`, the cause is null but `isCanceledOrFailed` would return true. What do you think?

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -879,6 +875,62 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
+	private static void closeNetworkResources(
+			ResultPartition[] producedPartitions,
+			InputGate[] inputGates,
+			boolean isCanceledOrFailed,
```
Could we rename it to just `isFailed`? Also, method `closeOrFailNetworkResources`.

zhijiangW replied:
Yes, we prefer this way.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -879,6 +875,62 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
+		closeNetworkResources(producedPartitions, inputGates, isCanceledOrFailed(), getFailureCause(), taskNameWithSubtask);
+	}
+
+	private static void closeNetworkResources(
```
This seems to be used only in one place, I would remove it.

zhijiangW commented:
@azagrebin  thanks for review again! I submitted a new commit for addressing above comments.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -1458,26 +1456,15 @@ private static AbstractInvokable loadAndInstantiateInvokable(
-				ResultPartition[] producedPartitions,
-				InputGate[] inputGates) {
 
+		TaskCanceler(Logger logger, Task task, AbstractInvokable invokable, Thread executer) {
```
Why are you introducing circular dependency here between `Task` and `TaskCanceler`? There are various reasons why this is bad, including: is it necessary to expose 27 public methods (including things like `startTaskThread()` or `run()`) to the `TaskCanceler`?

In various different places we are trying to get away from this pattern of passing `StreamingTask` everywhere.

zhijiangW replied:
The motivation is for reusing the `closeOrFailNetworkResources` and avoid static method and pass the arrays of `ResultPartition` and `InputGate` explicitly.  The previous `AbstractInvokable` could also be replaced and got from new `Task` parameter.

I think the previous introduced `AbstractInvokable` here is also not a good way considering exposing more public methods besides `AbstractInvokable#cancel()`. Comparing with `Task` parameter, we might add more public methods to do so. I agree with reverting this change to pass the previous specific three parameters here even though the `closeOrFailNetworkResources` might seem ugly.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -882,6 +878,47 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
+	 * @param isFailed true if the task has failed.
+	 * @param cause the exception that caused the task to fail, or null, if the task has not failed.
+	 */
+	private void closeOrFailNetworkResources(boolean isFailed, @Nullable Throwable cause) {
```
Thanks for de duplicating this logic already ( :) ), but you could also go one step further and maybe extract this logic to something like `TaskCloser` class and de-duplicate/re-use it in `TaskCanceller` as well?

zhijiangW replied:
The introduced `TaskCloser` seems better, agree with it.

azagrebin replied:
I would suggest to pass then `Runnable closeNetworkResources` to `TaskCanceler` constructor because it is still task concern to manage partitions/gates and we can avoid exploding methods with parameters. `TaskCanceler` does not really closes the whole task, just `interrupt`s network resources as well.

Also, just thought, that `TaskCanceler` actually does not need `isFailed` case.
@zhijiangW maybe, we could simplify `closeOrFailNetworkResources` and do only closing there. If we remove `close` from `partition.fail`, we could move conditional `partition.fail` to loop in `releaseNetworkResources`.

pnowojski replied:
Oh... ok, sorry by doing quick review I missed one that you have already deduplicated `TaskCanceler` code 😳 I thought that those are two separate issues that I'm commenting on, but clearly this is just one thing, so responding to both threads here in single comment:

Yes, passing `invokable` is also not the pretties thing, but it looks like this issue is beside the scope of this PR, right?

I haven't thought it through but passing `Runnable closeNetworkResources` seems fine to me as well. It has a drawback of being more or less the same thing, but with more vague name/type in an exchange of less overhead code. With simple class `TaskCloser` we could better specify concurrency contracts (`@ThreadSafe`) etc, but I think I would be fine both way.

zhijiangW commented:
Thanks for further reviews again @azagrebin @pnowojski 
I have squashed the commits for addressing above comments, because the previous first hotfix commit is not valid any more.

azagrebin commented:
LGTM 👍
